plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts

```diff
@@ -0,0 +1,261 @@
+ * limitations under the License.
+ */
+
+import {
+  AuthService,
+  CacheService,
+  DiscoveryService,
+  LoggerService,
+} from '@backstage/backend-plugin-api';
+import { CatalogClient } from '@backstage/catalog-client';
+import { Entity, UserEntity, parseEntityRef } from '@backstage/catalog-model';
+import { Config } from '@backstage/config';
+import { Notification } from '@backstage/plugin-notifications-common';
+import {
+  NotificationProcessor,
+  NotificationSendOptions,
+} from '@backstage/plugin-notifications-node';
+import { ANNOTATION_NOTIFICATION_SLACK_CHANNEL_ID } from './constants';
+
+const DEFAULT_CACHE_TTL = 1000 * 60 * 60; // 1 hour
+
+export class SlackNotificationProcessor implements NotificationProcessor {
+  private readonly logger: LoggerService;
+  private readonly directWebhookUrl?: string;
+  private readonly channelWebhookUrl?: string;
+  private readonly cache: CacheService;
+  private readonly ttl: number;
+  private readonly catalogClient: CatalogClient;
+  private readonly auth: AuthService;
+
+  static fromConfig(
+    config: Config,
+    options: {
+      auth: AuthService;
+      discovery: DiscoveryService;
+      logger: LoggerService;
+      cache: CacheService;
+      catalogClient?: CatalogClient; // used for testing
+    },
+  ): SlackNotificationProcessor[] {
+    const slackConfig =
+      config.getOptionalConfigArray('notifications.processors.slack') ?? [];
+    return slackConfig.map(c => {
+      const directWebhookUrl = c.getOptionalString('directWebhookUrl');
+      const channelWebhookUrl = c.getOptionalString('channelWebhookUrl');
+      const ttl = c.getOptionalNumber('cache.ttl') ?? DEFAULT_CACHE_TTL;
+      return new SlackNotificationProcessor({
+        directWebhookUrl,
+        channelWebhookUrl,
+        ttl,
+        ...options,
+      });
+    });
+  }
+
+  private constructor(options: {
+    auth: AuthService;
+    discovery: DiscoveryService;
+    logger: LoggerService;
+    directWebhookUrl?: string;
+    channelWebhookUrl?: string;
+    cache: CacheService;
+    ttl: number;
+    catalogClient?: CatalogClient;
+  }) {
+    const { auth, directWebhookUrl, channelWebhookUrl, logger, cache, ttl } =
+      options;
+    this.logger = logger;
+    this.directWebhookUrl = directWebhookUrl;
+    this.channelWebhookUrl = channelWebhookUrl;
+    this.cache = cache;
+    this.ttl = ttl;
+    this.auth = auth;
+    this.catalogClient =
+      options.catalogClient ??
+      new CatalogClient({
+        discoveryApi: options.discovery,
+      });
+  }
+  getName(): string {
+    return 'Slack';
+  }
+
+  async processOptions(
+    options: NotificationSendOptions,
+  ): Promise<NotificationSendOptions> {
+    if (!this.channelWebhookUrl || options.recipients.type !== 'entity') {
+      return options;
+    }
+
+    const entityRefs = Array.isArray(options.recipients.entityRef)
+      ? options.recipients.entityRef
+      : [options.recipients.entityRef];
```

Comment:
optional little thing: `[options.recipients.entityRef].flat()` is short and sweet! But probably a bit more wasteful